File: An2/MAP/MAP-Java/MAP-Lab8/src/Repository/Repository.java
package Repository;

import Model.HasID;
import Model.Comparable;
import Utils.MyException;

import java.io.*;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Created with IntelliJ IDEA.
 * User: mihai
 * Date: 12/4/13
 * Time: 11:47 PM
 * To change this template use File | Settings | File Templates.
 */
public class Repository<T extends HasID> {

    private Map<Integer, T> elements = new LinkedHashMap<Integer, T>();

    public Repository(Map<Integer, T> map){
        elements = map;
    }

    public Repository(){

    }

    public void addElement(T element){
        elements.put(element.getID(), element);
    }

    public void removeElement(T element){
        elements.remove(element.getID());
    }

    public T getTopElement(){
        for (T elem : elements.values())
            return elem;
        return null;
    }

    public int getNoOfElements(){
        return elements.size();
    }

    public Map<Integer, T> getAllElements(){
        Map<Integer, T> temp = new LinkedHashMap<Integer, T>();
        temp.putAll(elements);
        return temp;
    }

    public void writeToFile(String fileName) throws MyException {
        try {
            FileWriter out = new FileWriter(fileName);
            //Map<Integer, T> temp = getAllElements();
            String className;
            for (T elem : elements.values()){
                className = elem.getClass().toString().split("\\.")[1];
                out.write(className);
                out.write(" ");
                out.write(elem.toString());
                out.write("\n");
            }
            out.close();
        } catch (IOException ex){
            throw new MyException("Error writing to file!");
        }
    }

    public void serializeDataToFile(String fileName){
        try{
            FileOutputStream fileOut = new FileOutputStream(fileName);
            ObjectOutputStream out = new ObjectOutputStream(fileOut);
            out.writeObject(this.elements);
            out.close();
            fileOut.close();
        }catch (IOException ex){
            ex.printStackTrace();
        }
    }

    public void deserializeDataFromFile(String fileName){
        try{
            FileInputStream fileIn = new FileInputStream(fileName);
            ObjectInputStream in = new ObjectInputStream(fileIn);
            Map<Integer, T> map = (Map<Integer, T>) in.readObject();
            this.elements = map;
            in.close();
            fileIn.close();
        }catch (IOException ex){
            ex.printStackTrace();
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
    }

    public static<T extends Comparable<T>> int noOfElementsGreaterThan(Map<Integer,T> map, T elem){
        Map<Integer, T> temp = null;
        temp.putAll(map);
        int result = 0;

        for (T comparableElem : temp.values())
            if (comparableElem.isGreaterThan(elem))
                result++;

        return result;
    }

}

File: An2/Sem1/MAP/MAP-C#/MAP-Lab8/MAP-Lab8/Repository/Repository.cs
using MAP_Lab8.Model;
using MAP_Lab8.Utils;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading.Tasks;



namespace MAP_Lab8.Repository
{
    class Repository<T> where T : HasID
    {
        private IDictionary<int, T> elements = new Dictionary<int, T>();

        public Repository()
        { }

        public Repository(IDictionary<int, T> adt)
        {
            this.elements = adt;
        }

        public void addElement(T element)
        {
            this.elements.Add(element.getID(), element);
        }

        public void removeElement(T element)
        {
            this.elements.Remove(element.getID());
        }

        public T getTopElement()
        {
            foreach (T element in this.elements.Values)
                return element;
            return default(T);
        }

        public int getNoOfElements()
        {
            return this.elements.Count();
        }

        public IDictionary<int, T> getAllElements()
        {
            return new Dictionary<int, T>(this.elements);
        }

        public void writeToFile(String fileName)
        {
            try
            {
                StreamWriter writer = new StreamWriter(fileName);
                IDictionary<int, T> temp = this.getAllElements();
                String className;
                foreach (T element in temp.Values)
                {
                    className = element.GetType().Name;
                    writer.WriteLine(className + " " + element.ToString());
                }
                writer.Close();
            }
            catch (IOException)
            {
                throw new MyException("Problems in writing to file!");
            }
        }

        public void replaceContent(IDictionary<int, T> adt)
        {
            this.elements.Clear();
            this.elements = adt;
        }

        public void serializeDataToFile(string fileName)
        {
            Stream fileStream = File.Create(fileName);
            BinaryFormatter serializer = new BinaryFormatter();
            serializer.Serialize(fileStream, this.elements);
            fileStream.Close();
        }

        public void deserializeDataFromFile(string fileName)
        {
            if (File.Exists(fileName))
            {
                Stream fileStream = File.OpenRead(fileName);
                BinaryFormatter deserializer = new BinaryFormatter();
                IDictionary<int, T> adt = (IDictionary<int, T>)deserializer.Deserialize(fileStream);
                this.elements = adt;
                fileStream.Close();
            }
        }

        public static int noOfElementsGreaterThan<T>(IDictionary<int, T> adt, T elem) where T : Model.Comparable<T>
        {
            int result = 0;
            IDictionary<int, T> temp = new Dictionary<int, T>(adt);

            foreach (T element in temp.Values)
                if (element.isGreaterThan(elem))
                    result++;

            return result;
        }

    }
}
